## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#ccb97efc-0e56-4e6d-aedf-2b91b35a3da5) |
| mint-erasure.sh | :x: [more...](#ea152607-4f08-4652-a191-b7babee3a518) |
| mint-dist-erasure.sh | :x: [more...](#20ca64eb-9de4-4422-818d-2dc3cd1b9c29) |
| mint-zoned.sh | :x: [more...](#9f3cab1f-55a5-4534-987b-07617ae107f6) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#41194876-7e5a-4429-9a14-7a8c23bc8863) |
<a name="ea152607-4f08-4652-a191-b7babee3a518"></a>
### 11822-b528e6e/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32683
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a731a90a7996:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-56jdfzs5y9g0sl",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 134,
  "error": "BucketNotEmpty: The bucket you tried to delete is not empty\n\tstatus code: 409, request id: 166D5355B99CFE76, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 2 seconds
(2/15) Running aws-sdk-php tests ... FAILED in 0 seconds
{
  "name": "aws-sdk-php",
  "function": "listObjects ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-6052ce2e86088",
    "Object": "obj1"
  },
  "duration": "156",
  "status": "FAIL",
  "error": "Error executing \"CopyObject\" on \"http://minio-dev4.minio.io:32683/aws-sdk-php-6052ce2e86088/obj1-copy-0\"; AWS HTTP error: Server error: `PUT http://minio-dev4.minio.io:32683/aws-sdk-php-6052ce2e86088/obj1-copy-0` resulted in a `500 Internal Server Error` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InternalError</Code><Message>We encountered an internal error, pleas (truncated...)\n InternalError (server): We encountered an internal error, please try again.: cause(disk does not support O_DIRECT) - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InternalError</Code><Message>We encountered an internal error, please try again.: cause(disk does not support O_DIRECT)</Message><Key>obj1-copy-0</Key><BucketName>aws-sdk-php-6052ce2e86088</BucketName><Resource>/aws-sdk-php-6052ce2e86088/obj1-copy-0</Resource><RequestId>166D53563226EF76</RequestId><HostId>4e2f676f-cdab-4dfb-888e-17b90dea7454</HostId></Error>"
}
(2/15) Running aws-sdk-ruby tests ... FAILED in 1 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-da093a7bc71a"
  },
  "duration": 364.37,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-da093a7bc71a', The bucket you tried to delete is not empty",
  "status": "FAIL"
}
(2/15) Running awscli tests ... FAILED in 14 seconds
{
  "name": "awscli",
  "duration": 10244,
  "function": "aws --endpoint-url http://minio-dev4.minio.io:32683 s3api get-object --bucket awscli-mint-test-bucket-19475 --key datafile-1-kB /tmp/datafile-1-kB\n",
  "status": "FAIL",
  "error": "An error occurred (InternalError) when calling the GetObject operation (reached max retries: 4): We encountered an internal error, please try again.: cause(disk does not support O_DIRECT)"
}
(2/15) Running healthcheck tests ... done in 0 seconds
(3/15) Running mc tests ... FAILED in 2 minutes and 11 seconds
{
  "name": "mc",
  "duration": "69234",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-11857 --quiet --no-color rm myminio/mc-test-bucket-10727/mc-test-object-26816  >>> \nmc: <ERROR> Failed to remove `myminio/mc-test-bucket-10727/mc-test-object-26816`. 500 Internal Server Error"
}
(3/15) Running minio-dotnet tests ... FAILED in 27 seconds
{
  "name": "minio-dotnet : CopyObject_Test1",
  "function": "Task<CopyObjectResult> CopyObjectAsync(string bucketName, string objectName, string destBucketName, string destObjectName = null, CopyConditions copyConditions = null, CancellationToken cancellationToken = default(CancellationToken))",
  "args": {
    "bucketName": "miniodotnetkstd6eqp5oxnf2r",
    "objectName": "%*a)a$d&a{",
    "destBucketName": "miniodotnet8hb1x7vwcqrk9zg",
    "destObjectName": "miniodotnet229f6k76yb",
    "data": "1KB",
    "size": "1KB"
  },
  "duration": 207,
  "status": "FAIL",
  "alert": "",
  "message": "MinIO API responded with message=Unsuccessful response from server without XML error: . Status code=InternalServerError, response=, content=",
  "error": "Minio.Exceptions.InternalClientException: MinIO API responded with message=Unsuccessful response from server without XML error: . Status code=InternalServerError, response=, content=\n   at Minio.MinioClient.ParseErrorNoContent(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 447\n   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 431\n   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 605\n   at Minio.MinioClient.ExecuteTaskCoreAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 405\n   at Minio.MinioClient.StatObjectAsync(String bucketName, String objectName, ServerSideEncryption sse, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 820\n   at Minio.MinioClient.CopyObjectAsync(String bucketName, String objectName, String destBucketName, String destObjectName, CopyConditions copyConditions, Dictionary`2 metadata, ServerSideEncryption sseSrc, ServerSideEncryption sseDest, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 945\n   at Minio.Functional.Tests.FunctionalTest.CopyObject_Test1(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 930"
}
(3/15) Running minio-go tests ... FAILED in 5 seconds
{
  "args": {
    "bucketName": "minio-go-test-rhqftb90i5i6eb9i",
    "objectName": "vnmi2x3mcn6esujwecym6ran3yogt5",
    "region": "eu-west-1"
  },
  "duration": 4248,
  "error": "500 Internal Server Error",
  "function": "MakeBucket(bucketName, region)",
  "message": "Stat failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(3/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "setBucketVersioning()",
  "duration": 85,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InternalError, message = We encountered an internal error, please try again.: cause(disk does not support O_DIRECT), bucketName = minio-java-test-106j1i2, objectName = null, resource = /minio-java-test-106j1i2, requestId = 166D538012EC9E19, hostId = 4e2f676f-cdab-4dfb-888e-17b90dea7454)\nrequest={method=PUT, url=http://minio-dev4.minio.io:32683/minio-java-test-106j1i2?versioning=, headers=Host: minio-dev4.minio.io:32683\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: QQFYoy/mRYV9PGZUfFi0Bw==\nx-amz-content-sha256: 482d4e510b5cdcf0bbf3a044e82832dc711f4f28757a6c06eb21758282a73ef7\nx-amz-date: 20210318T035410Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210318/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=500, headers=Accept-Ranges: bytes\nContent-Length: 374\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 166D538012EC9E19\nX-Xss-Protection: 1; mode=block\nDate: Thu, 18 Mar 2021 03:54:10 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.setBucketVersioning(MinioClient.java:2313), FunctionalTest.setBucketVersioning(FunctionalTest.java:606), FunctionalTest.runBucketTests(FunctionalTest.java:3713), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(3/15) Running minio-js tests ... FAILED in 2 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 8,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(3/15) Running minio-py tests ... FAILED in 6 seconds
{
  "name": "minio-py:test_fput_object_small_file",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-29c11fc7-0234-4125-aeb0-52665c450e2e",
    "object_name": "d3a51bdf-7f70-43ea-8e31-4b0909520347-f",
    "file_path": "/mint/data/datafile-1-MB",
    "metadata": {
      "x-amz-storage-class": "STANDARD_IA"
    }
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-29c11fc7-0234-4125-aeb0-52665c450e2e, request_id: 166D5381F42CA72B, host_id: 4e2f676f-cdab-4dfb-888e-17b90dea7454, bucket_name: minio-py-test-29c11fc7-0234-4125-aeb0-52665c450e2e",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 310, in _test_fput_object\n    _CLIENT.stat_object(bucket_name, object_name, ssec=sse)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1894, in stat_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 278, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 836, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 573, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev4.minio.io', port=32683): Max retries exceeded with url: /minio-py-test-29c11fc7-0234-4125-aeb0-52665c450e2e/d3a51bdf-7f70-43ea-8e31-4b0909520347-f (Caused by ResponseError('too many 500 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 334, in test_fput_object_small_file\n    _test_fput_object(bucket_name, object_name, _TEST_FILE, metadata, sse)\n  File \"/mint/run/core/minio-py/tests.py\", line 313, in _test_fput_object\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-29c11fc7-0234-4125-aeb0-52665c450e2e, request_id: 166D5381F42CA72B, host_id: 4e2f676f-cdab-4dfb-888e-17b90dea7454, bucket_name: minio-py-test-29c11fc7-0234-4125-aeb0-52665c450e2e\n",
  "duration": 6448
}
(3/15) Running s3cmd tests ... FAILED in 52 seconds
{
  "name": "s3cmd",
  "duration": "49153",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: Upload failed: /s3cmd-test-object-18571?partNumber=1&uploadId=e9d6d5a7-b12e-4403-af6f-a1f7ecb8e73d (500 (InternalError): We encountered an internal error, please try again.: cause(disk does not support O_DIRECT))\nWARNING: Waiting 3 sec...\nWARNING: Upload failed: /s3cmd-test-object-18571?partNumber=1&uploadId=e9d6d5a7-b12e-4403-af6f-a1f7ecb8e73d (500 (InternalError): We encountered an internal error, please try again.: cause(disk does not support O_DIRECT))\nWARNING: Waiting 6 sec...\nWARNING: Upload failed: /s3cmd-test-object-18571?partNumber=1&uploadId=e9d6d5a7-b12e-4403-af6f-a1f7ecb8e73d (500 (InternalError): We encountered an internal error, please try again.: cause(disk does not support O_DIRECT))\nWARNING: Waiting 9 sec...\nWARNING: Upload failed: /s3cmd-test-object-18571?partNumber=1&uploadId=e9d6d5a7-b12e-4403-af6f-a1f7ecb8e73d (500 (InternalError): We encountered an internal error, please try again.: cause(disk does not support O_DIRECT))\nWARNING: Waiting 12 sec...\nWARNING: Upload failed: /s3cmd-test-object-18571?partNumber=1&uploadId=e9d6d5a7-b12e-4403-af6f-a1f7ecb8e73d (500 (InternalError): We encountered an internal error, please try again.: cause(disk does not support O_DIRECT))\nWARNING: Waiting 15 sec...\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-7376/s3cmd-test-object-18571 e9d6d5a7-b12e-4403-af6f-a1f7ecb8e73d\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id e9d6d5a7-b12e-4403-af6f-a1f7ecb8e73d put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: Upload failed for: /s3cmd-test-object-18571?partNumber=1&uploadId=e9d6d5a7-b12e-4403-af6f-a1f7ecb8e73d)"
}
(3/15) Running s3select tests ... FAILED in 1 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, request)",
  "args": {
    "bucket_name": "s3select-test-765bbbd5-f600-407d-9929-3cb720c64069"
  },
  "duration": 200,
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-765bbbd5-f600-407d-9929-3cb720c64069, request_id: 166D538E24CAC78C, host_id: 4e2f676f-cdab-4dfb-888e-17b90dea7454, bucket_name: s3select-test-765bbbd5-f600-407d-9929-3cb720c64069",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/s3select/csv.py\", line 39, in test_sql_api\n    data = client.select_object_content(bucket_name, object_name, sql_opts)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 576, in select_object_content\n    preload_content=False,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: InternalError, message: We encountered an internal error, please try again.: cause(disk does not support O_DIRECT), resource: /s3select-test-765bbbd5-f600-407d-9929-3cb720c64069/038e3c7c-2bb9-4159-9d9d-8d2fde3e8347, request_id: 166D538E22EAFEA2, host_id: 4e2f676f-cdab-4dfb-888e-17b90dea7454, bucket_name: s3select-test-765bbbd5-f600-407d-9929-3cb720c64069, object_name: 038e3c7c-2bb9-4159-9d9d-8d2fde3e8347\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/s3select/csv.py\", line 118, in test_csv_input_custom_quote_char\n    data, sql_opts, expected_output)\n  File \"/mint/run/core/s3select/csv.py\", line 48, in test_sql_api\n    'Test {} unexpectedly failed with: {}'.format(test_name, select_err))\nValueError: Test test_1 unexpectedly failed with: S3 operation failed; code: InternalError, message: We encountered an internal error, please try again.: cause(disk does not support O_DIRECT), resource: /s3select-test-765bbbd5-f600-407d-9929-3cb720c64069/038e3c7c-2bb9-4159-9d9d-8d2fde3e8347, request_id: 166D538E22EAFEA2, host_id: 4e2f676f-cdab-4dfb-888e-17b90dea7454, bucket_name: s3select-test-765bbbd5-f600-407d-9929-3cb720c64069, object_name: 038e3c7c-2bb9-4159-9d9d-8d2fde3e8347\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 53, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 120, in test_csv_input_custom_quote_char\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /s3select-test-765bbbd5-f600-407d-9929-3cb720c64069, request_id: 166D538E24CAC78C, host_id: 4e2f676f-cdab-4dfb-888e-17b90dea7454, bucket_name: s3select-test-765bbbd5-f600-407d-9929-3cb720c64069\n",
  "status": "FAIL"
}
(3/15) Running security tests ... done in 0 seconds

Executed 3 out of 15 tests successfully.
```
<a name="41194876-7e5a-4429-9a14-7a8c23bc8863"></a>
### 11822-b528e6e/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31338
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 49aff555c81e:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ybxyljfqb4sexr",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 365354,
  "error": "SlowDown: Please reduce your request\n\tstatus code: 503, request id: 166D53AACC2204CB, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 1 seconds
(2/15) Running aws-sdk-php tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-php",
  "function": "listMultipartUploads ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-6052cf9bc3101",
    "Object": "obj1"
  },
  "duration": "363",
  "status": "FAIL",
  "error": "Error executing \"UploadPartCopy\" on \"http://minio-dev6.minio.io:31338/aws-sdk-php-6052cf9bc3101/obj1-copy?uploadId=c4654eec-3131-4781-a27d-e77780e0c3b6&partNumber=1\"; AWS HTTP error: Server error: `PUT http://minio-dev6.minio.io:31338/aws-sdk-php-6052cf9bc3101/obj1-copy?uploadId=c4654eec-3131-4781-a27d-e77780e0c3b6&partNumber=1` resulted in a `503 Service Unavailable` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>SlowDown</Code><Message>Please reduce your request</Message><Key>obj (truncated...)\n SlowDown (server): Please reduce your request - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>SlowDown</Code><Message>Please reduce your request</Message><Key>obj1-copy</Key><BucketName>aws-sdk-php-6052cf9bc3101</BucketName><Resource>/aws-sdk-php-6052cf9bc3101/obj1-copy</Resource><RequestId>166D53AB93A2C2E4</RequestId><HostId>27112822-657a-42f2-bf35-ca57657341e0</HostId></Error>"
}
(2/15) Running aws-sdk-ruby tests ... FAILED in 5 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-5dc0d474c466"
  },
  "duration": 3545.62,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-5dc0d474c466', Please reduce your request",
  "status": "FAIL"
}
(2/15) Running awscli tests ... FAILED in 18 seconds
{
  "name": "awscli",
  "duration": 14032,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-11400/datafile-1-kBremove_bucket failed: s3://awscli-mint-test-bucket-11400 An error occurred (SlowDown) when calling the DeleteBucket operation (reached max retries: 4): Please reduce your request"
}
(2/15) Running healthcheck tests ... done in 1 seconds
(3/15) Running mc tests ... FAILED in 3 seconds
{
  "name": "mc",
  "duration": "590",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-26761 --quiet --no-color rm myminio/mc-test-bucket-10381/mc-test-object-18852  >>> \nRemoving `myminio/mc-test-bucket-10381/mc-test-object-18852`.\nmc: <ERROR> Failed to remove `myminio/mc-test-bucket-10381/mc-test-object-18852`. Please reduce your request"
}
(3/15) Running minio-dotnet tests ... done in 37 seconds
(4/15) Running minio-go tests ... FAILED in 39 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-htcxr0zafq2hivhf",
      "Object": "dstObject",
      "Encryption": null,
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "sourceList": [
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-htcxr0zafq2hivhf",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": true,
        "Start": 0,
        "End": 0,
        "Encryption": null
      }
    ]
  },
  "duration": 3734,
  "error": "Please reduce your request",
  "function": "ComposeObject(destination, sourceList)",
  "message": "ComposeObject failed",
  "name": "minio-go: testComposeMultipleSources",
  "status": "FAIL"
}
(4/15) Running minio-java tests ... FAILED in 10 seconds
(4/15) Running minio-js tests ... FAILED in 2 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 7,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(4/15) Running minio-py tests ... 
Aborting Mint...
```
<a name="9f3cab1f-55a5-4534-987b-07617ae107f6"></a>
### 11822-b528e6e/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:30473
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 14709e8e587a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 2 minutes and 2 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-fjtiz5sds139gz",
    "objectName1": "object1.csv",
    "objectName2": "object2.csv"
  },
  "duration": 121145,
  "error": "InternalError: We encountered an internal error, please try again.: cause(disk does not support O_DIRECT)\n\tstatus code: 500, request id: 166D5372095EA44F, host id: ",
  "function": "testSelectObject",
  "message": "AWS SDK Go Select object failed InternalError: We encountered an internal error, please try again.: cause(disk does not support O_DIRECT)\n\tstatus code: 500, request id: 166D5372095EA44F, host id: ",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 4 seconds
(2/15) Running aws-sdk-php tests ... FAILED in 3 seconds
{
  "name": "aws-sdk-php",
  "function": "anonDeleteObjects ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-6052ceaa3cc6c",
    "Object": "obj1"
  },
  "duration": "192",
  "status": "FAIL",
  "error": "Error executing \"CopyObject\" on \"http://minio-c3.minio.io:30473/aws-sdk-php-6052ceaa3cc6c/obj1-copy0\"; AWS HTTP error: Server error: `PUT http://minio-c3.minio.io:30473/aws-sdk-php-6052ceaa3cc6c/obj1-copy0` resulted in a `500 Internal Server Error` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InternalError</Code><Message>We encountered an internal error, pleas (truncated...)\n InternalError (server): We encountered an internal error, please try again.: cause(disk does not support O_DIRECT) - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InternalError</Code><Message>We encountered an internal error, please try again.: cause(disk does not support O_DIRECT)</Message><Key>obj1-copy0</Key><BucketName>aws-sdk-php-6052ceaa3cc6c</BucketName><Resource>/aws-sdk-php-6052ceaa3cc6c/obj1-copy0</Resource><RequestId>166D5373BA973CE9</RequestId><HostId>423f5015-755b-4fe8-b5ef-6815d34025e5</HostId></Error>"
}
(2/15) Running aws-sdk-ruby tests ... FAILED in 8 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-5ff20e85d761"
  },
  "duration": 2483.29,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-5ff20e85d761', The bucket you tried to delete is not empty",
  "status": "FAIL"
}
(2/15) Running awscli tests ... FAILED in 21 seconds
{
  "name": "awscli",
  "duration": 13736,
  "function": "aws --endpoint-url http://minio-c3.minio.io:30473 s3api put-object --bucket awscli-mint-test-bucket-1508 --key prefix/directory/\n",
  "status": "FAIL",
  "error": "An error occurred (InternalError) when calling the PutObject operation (reached max retries: 4): We encountered an internal error, please try again.: cause(disk does not support O_DIRECT)"
}
(2/15) Running healthcheck tests ... done in 0 seconds
(3/15) Running mc tests ... FAILED in 2 minutes and 13 seconds
{
  "name": "mc",
  "duration": "68693",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-16603 --quiet --no-color rm myminio/mc-test-bucket-5296/mc-test-object-26570  >>> \nmc: <ERROR> Failed to remove `myminio/mc-test-bucket-5296/mc-test-object-26570`. 503 Service Unavailable"
}
(3/15) Running minio-dotnet tests ... done in 1 minutes and 18 seconds
(4/15) Running minio-go tests ... FAILED in 5 seconds
{
  "args": {
    "bucketName": "minio-go-test-n491uefj6y9b43eg",
    "objectName": "fbxzj4tuc654o39gq6pj6no5fficbc",
    "region": "eu-west-1"
  },
  "duration": 5277,
  "error": "503 Service Unavailable",
  "function": "MakeBucket(bucketName, region)",
  "message": "Stat failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(4/15) Running minio-java tests ... FAILED in 15 seconds
{
  "name": "minio-java",
  "function": "setBucketVersioning()",
  "duration": 2128,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = InternalError, message = We encountered an internal error, please try again.: cause(disk does not support O_DIRECT), bucketName = minio-java-test-15iinpg, objectName = null, resource = /minio-java-test-15iinpg, requestId = 166D53B016E6ADD4, hostId = 423f5015-755b-4fe8-b5ef-6815d34025e5)\nrequest={method=PUT, url=http://minio-c3.minio.io:30473/minio-java-test-15iinpg?versioning=, headers=Host: minio-c3.minio.io:30473\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: QQFYoy/mRYV9PGZUfFi0Bw==\nx-amz-content-sha256: 482d4e510b5cdcf0bbf3a044e82832dc711f4f28757a6c06eb21758282a73ef7\nx-amz-date: 20210318T035737Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210318/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=500, headers=Accept-Ranges: bytes\nContent-Length: 374\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 166D53B016E6ADD4\nX-Xss-Protection: 1; mode=block\nDate: Thu, 18 Mar 2021 03:57:37 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.setBucketVersioning(MinioClient.java:2313), FunctionalTest.setBucketVersioning(FunctionalTest.java:606), FunctionalTest.runBucketTests(FunctionalTest.java:3713), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(4/15) Running minio-js tests ... FAILED in 43 seconds
{
  "name": "minio-js",
  "function": "makeBucket(bucketName, region)",
  "args": "bucketName:minio-js-test-8dadc144-b4b4-4b26-9b15-c71cfcb177b3-region, region:us-east-1",
  "duration": 40098,
  "status": "FAIL",
  "error": "S3Error: Your previous request to create the named bucket succeeded and you already own it. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(4/15) Running minio-py tests ... FAILED in 15 seconds
{
  "name": "minio-py:test_copy_object_etag_match",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-2ac82be8-43b5-41ad-be1b-01e6d8ae0a70",
    "object_source": "9b8cf2f6-b3fa-4611-8c37-d1b802986c1e-source",
    "object_name": "9b8cf2f6-b3fa-4611-8c37-d1b802986c1e-copy"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-2ac82be8-43b5-41ad-be1b-01e6d8ae0a70, request_id: 166D53BDA69B65FE, host_id: 423f5015-755b-4fe8-b5ef-6815d34025e5, bucket_name: minio-py-test-2ac82be8-43b5-41ad-be1b-01e6d8ae0a70",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 524, in test_copy_object_etag_match\n    bucket_name, object_copy, CopySource(bucket_name, object_source),\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1265, in copy_object\n    ssec=source.ssec,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1894, in stat_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 278, in _url_open\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 375, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 859, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 836, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 573, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-c3.minio.io', port=30473): Max retries exceeded with url: /minio-py-test-2ac82be8-43b5-41ad-be1b-01e6d8ae0a70/9b8cf2f6-b3fa-4611-8c37-d1b802986c1e-source (Caused by ResponseError('too many 500 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 536, in test_copy_object_etag_match\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-2ac82be8-43b5-41ad-be1b-01e6d8ae0a70, request_id: 166D53BDA69B65FE, host_id: 423f5015-755b-4fe8-b5ef-6815d34025e5, bucket_name: minio-py-test-2ac82be8-43b5-41ad-be1b-01e6d8ae0a70\n",
  "duration": 8247
}
(4/15) Running s3cmd tests ... FAILED in 27 seconds
{
  "name": "s3cmd",
  "duration": "525",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "ERROR: S3 error: 400 (InvalidRequest): Bucket is missing ObjectLockConfiguration"
}
(4/15) Running s3select tests ... 
Aborting Mint...
```
<a name="20ca64eb-9de4-4422-818d-2dc3cd1b9c29"></a>
### 11822-b528e6e/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30574
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9dc5ef60e1e7:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-u9ty6mid0p2tol",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 365869,
  "error": "SlowDown: Please reduce your request\n\tstatus code: 503, request id: 166D53AAD491350C, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 1 seconds
(2/15) Running aws-sdk-php tests ... FAILED in 2 seconds
{
  "name": "aws-sdk-php",
  "function": "listMultipartUploads ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-6052cf9bf2a51",
    "Object": "obj1"
  },
  "duration": "348",
  "status": "FAIL",
  "error": "Error executing \"UploadPartCopy\" on \"http://minio-dev6.minio.io:30574/aws-sdk-php-6052cf9bf2a51/obj1-copy?uploadId=38fd997e-2f7c-4a42-8961-7231a9caa3c9&partNumber=1\"; AWS HTTP error: Server error: `PUT http://minio-dev6.minio.io:30574/aws-sdk-php-6052cf9bf2a51/obj1-copy?uploadId=38fd997e-2f7c-4a42-8961-7231a9caa3c9&partNumber=1` resulted in a `503 Service Unavailable` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>SlowDown</Code><Message>Please reduce your request</Message><Key>obj (truncated...)\n SlowDown (server): Please reduce your request - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>SlowDown</Code><Message>Please reduce your request</Message><Key>obj1-copy</Key><BucketName>aws-sdk-php-6052cf9bf2a51</BucketName><Resource>/aws-sdk-php-6052cf9bf2a51/obj1-copy</Resource><RequestId>166D53AB75324D78</RequestId><HostId>6cf57253-e066-45b5-8922-b98b985f9a65</HostId></Error>"
}
(2/15) Running aws-sdk-ruby tests ... FAILED in 4 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-28fb5dd5cef9"
  },
  "duration": 2646.99,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-28fb5dd5cef9', Please reduce your request",
  "status": "FAIL"
}
(2/15) Running awscli tests ... FAILED in 19 seconds
{
  "name": "awscli",
  "duration": 15045,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-12572/datafile-1-kBremove_bucket failed: s3://awscli-mint-test-bucket-12572 An error occurred (SlowDown) when calling the DeleteBucket operation (reached max retries: 4): Please reduce your request"
}
(2/15) Running healthcheck tests ... done in 0 seconds
(3/15) Running mc tests ... FAILED in 2 seconds
{
  "name": "mc",
  "duration": "383",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-6152 --quiet --no-color rm myminio/mc-test-bucket-13775/mc-test-object-3333  >>> \nRemoving `myminio/mc-test-bucket-13775/mc-test-object-3333`.\nmc: <ERROR> Failed to remove `myminio/mc-test-bucket-13775/mc-test-object-3333`. Please reduce your request"
}
(3/15) Running minio-dotnet tests ... done in 37 seconds
(4/15) Running minio-go tests ... FAILED in 43 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
      "Object": "dstObject",
      "Encryption": null,
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "sourceList": [
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-u3p9hyg9sfvizyoi",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": true,
        "Start": 0,
        "End": 0,
        "Encryption": null
      }
    ]
  },
  "duration": 3969,
  "error": "Please reduce your request",
  "function": "ComposeObject(destination, sourceList)",
  "message": "ComposeObject failed",
  "name": "minio-go: testComposeMultipleSources",
  "status": "FAIL"
}
(4/15) Running minio-java tests ... FAILED in 8 seconds
(4/15) Running minio-js tests ... FAILED in 3 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 10,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(4/15) Running minio-py tests ... 
Aborting Mint...
```
<a name="ccb97efc-0e56-4e6d-aedf-2b91b35a3da5"></a>
### 11822-b528e6e/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31193
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 51215399691d:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-k53qowz0dbjccb",
    "expiry": 60000000000,
    "objectName1": "testObject1",
    "objectName2": "testObject2"
  },
  "duration": 366004,
  "error": "SlowDown: Please reduce your request\n\tstatus code: 503, request id: 166D53AB218DBF8B, host id: ",
  "function": "testListObjects",
  "message": "AWS SDK Go DeleteBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 2 seconds
(2/15) Running aws-sdk-php tests ... FAILED in 1 seconds
{
  "name": "aws-sdk-php",
  "function": "listMultipartUploads ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-6052cf9d5ecff",
    "Object": "obj1"
  },
  "duration": "278",
  "status": "FAIL",
  "error": "Error executing \"UploadPartCopy\" on \"http://minio-dev7.minio.io:31193/aws-sdk-php-6052cf9d5ecff/obj1-copy?uploadId=2f451d8d-3ce9-4a74-a7e5-9cd31dc5839e&partNumber=1\"; AWS HTTP error: Server error: `PUT http://minio-dev7.minio.io:31193/aws-sdk-php-6052cf9d5ecff/obj1-copy?uploadId=2f451d8d-3ce9-4a74-a7e5-9cd31dc5839e&partNumber=1` resulted in a `503 Service Unavailable` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>SlowDown</Code><Message>Please reduce your request</Message><Key>obj (truncated...)\n SlowDown (server): Please reduce your request - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>SlowDown</Code><Message>Please reduce your request</Message><Key>obj1-copy</Key><BucketName>aws-sdk-php-6052cf9d5ecff</BucketName><Resource>/aws-sdk-php-6052cf9d5ecff/obj1-copy</Resource><RequestId>166D53ABC3E1ED62</RequestId><HostId>3783c4ef-33c8-470c-a234-f7fdb594619e</HostId></Error>"
}
(2/15) Running aws-sdk-ruby tests ... FAILED in 4 seconds
{
  "name": "aws-sdk-ruby",
  "function": "removeBucket(bucket_name)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-f3a938a2481d"
  },
  "duration": 2693.25,
  "error": "Failed to clean-up bucket 'aws-sdk-ruby-bucket-f3a938a2481d', Please reduce your request",
  "status": "FAIL"
}
(2/15) Running awscli tests ... FAILED in 26 seconds
{
  "name": "awscli",
  "duration": 21546,
  "function": "delete_bucket\n",
  "status": "FAIL",
  "error": "delete: s3://awscli-mint-test-bucket-13033/datafile-1-kBremove_bucket failed: s3://awscli-mint-test-bucket-13033 An error occurred (SlowDown) when calling the DeleteBucket operation (reached max retries: 4): Please reduce your request"
}
(2/15) Running healthcheck tests ... done in 0 seconds
(3/15) Running mc tests ... FAILED in 3 seconds
{
  "name": "mc",
  "duration": "652",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-8392 --quiet --no-color rm myminio/mc-test-bucket-23019/mc-test-object-4422  >>> \nRemoving `myminio/mc-test-bucket-23019/mc-test-object-4422`.\nmc: <ERROR> Failed to remove `myminio/mc-test-bucket-23019/mc-test-object-4422`. Please reduce your request"
}
(3/15) Running minio-dotnet tests ... done in 38 seconds
(4/15) Running minio-go tests ... FAILED in 44 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-zznh4q1leobm6g9c",
      "Object": "dstObject",
      "Encryption": null,
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "sourceList": [
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-zznh4q1leobm6g9c",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": true,
        "Start": 0,
        "End": 0,
        "Encryption": null
      }
    ]
  },
  "duration": 4753,
  "error": "Please reduce your request",
  "function": "ComposeObject(destination, sourceList)",
  "message": "ComposeObject failed",
  "name": "minio-go: testComposeMultipleSources",
  "status": "FAIL"
}
(4/15) Running minio-java tests ... FAILED in 8 seconds
(4/15) Running minio-js tests ... FAILED in 2 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 8,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(4/15) Running minio-py tests ... 
Aborting Mint...
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/bitrot-streaming.go
+++ b/cmd/bitrot-streaming.go
@@ -23,8 +23,11 @@ import (
 	"fmt"
 	"hash"
 	"io"
+	"time"
 
 	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/pkg/env"
+	xioutil "github.com/minio/minio/pkg/ioutil"
 )
 
 type errHashMismatch struct {
@@ -37,7 +40,7 @@ func (err *errHashMismatch) Error() string {
 
 // Calculates bitrot in chunks and writes the hash into the stream.
 type streamingBitrotWriter struct {
-	iow       *io.PipeWriter
+	iow       io.WriteCloser
 	h         hash.Hash
 	shardSize int64
 	canClose  chan struct{} // Needed to avoid race explained in Close() call.
@@ -70,19 +73,28 @@ func (b *streamingBitrotWriter) Close() error {
 	return err
 }
 
+var (
+	ioDeadline, _ = time.ParseDuration(env.Get("MINIO_IO_DEADLINE", ""))
+)
+
 // Returns streaming bitrot writer implementation.
-func newStreamingBitrotWriter(disk StorageAPI, volume, filePath string, length int64, algo BitrotAlgorithm, shardSize int64) io.WriteCloser {
+func newStreamingBitrotWriter(disk StorageAPI, volume, filePath string, length int64, algo BitrotAlgorithm, shardSize int64, heal bool) io.Writer {
 	r, w := io.Pipe()
 	h := algo.New()
-	bw := &streamingBitrotWriter{w, h, shardSize, make(chan struct{})}
+
+	var wc io.WriteCloser = w
+	if ioDeadline > 0 && !heal {
+		wc = xioutil.NewDeadlineWriter(w, ioDeadline)
+	}
+
+	bw := &streamingBitrotWriter{wc, h, shardSize, make(chan struct{})}
 	go func() {
 		totalFileSize := int64(-1) // For compressed objects length will be unknown (represented by length=-1)
 		if length != -1 {
 			bitrotSumsTotalSize := ceilFrac(length, shardSize) * int64(h.Size()) // Size used for storing bitrot checksums.
 			totalFileSize = bitrotSumsTotalSize + length
 		}
-		err := disk.CreateFile(context.TODO(), volume, filePath, totalFileSize, r)
-		r.CloseWithError(err)
+		r.CloseWithError(disk.CreateFile(context.TODO(), volume, filePath, totalFileSize, r))
 		close(bw.canClose)
 	}()
 	return bw

--- a/cmd/bitrot.go
+++ b/cmd/bitrot.go
@@ -96,9 +96,9 @@ func BitrotAlgorithmFromString(s string) (a BitrotAlgorithm) {
 	return
 }
 
-func newBitrotWriter(disk StorageAPI, volume, filePath string, length int64, algo BitrotAlgorithm, shardSize int64) io.Writer {
+func newBitrotWriter(disk StorageAPI, volume, filePath string, length int64, algo BitrotAlgorithm, shardSize int64, heal bool) io.Writer {
 	if algo == HighwayHash256S {
-		return newStreamingBitrotWriter(disk, volume, filePath, length, algo, shardSize)
+		return newStreamingBitrotWriter(disk, volume, filePath, length, algo, shardSize, heal)
 	}
 	return newWholeBitrotWriter(disk, volume, filePath, algo, shardSize)
 }

--- a/cmd/bitrot_test.go
+++ b/cmd/bitrot_test.go
@@ -41,7 +41,7 @@ func testBitrotReaderWriterAlgo(t *testing.T, bitrotAlgo BitrotAlgorithm) {
 
 	disk.MakeVol(context.Background(), volume)
 
-	writer := newBitrotWriter(disk, volume, filePath, 35, bitrotAlgo, 10)
+	writer := newBitrotWriter(disk, volume, filePath, 35, bitrotAlgo, 10, false)
 
 	_, err = writer.Write([]byte("aaaaaaaaaa"))
 	if err != nil {

--- a/cmd/erasure-decode_test.go
+++ b/cmd/erasure-decode_test.go
@@ -108,7 +108,8 @@ func TestErasureDecode(t *testing.T) {
 		buffer := make([]byte, test.blocksize, 2*test.blocksize)
 		writers := make([]io.Writer, len(disks))
 		for i, disk := range disks {
-			writers[i] = newBitrotWriter(disk, "testbucket", "object", erasure.ShardFileSize(test.data), writeAlgorithm, erasure.ShardSize())
+			writers[i] = newBitrotWriter(disk, "testbucket", "object",
+				erasure.ShardFileSize(test.data), writeAlgorithm, erasure.ShardSize(), false)
 		}
 		n, err := erasure.Encode(context.Background(), bytes.NewReader(data[:]), writers, buffer, erasure.dataBlocks+1)
 		closeBitrotWriters(writers)
@@ -234,7 +235,8 @@ func TestErasureDecodeRandomOffsetLength(t *testing.T) {
 		if disk == nil {
 			continue
 		}
-		writers[i] = newBitrotWriter(disk, "testbucket", "object", erasure.ShardFileSize(length), DefaultBitrotAlgorithm, erasure.ShardSize())
+		writers[i] = newBitrotWriter(disk, "testbucket", "object",
+			erasure.ShardFileSize(length), DefaultBitrotAlgorithm, erasure.ShardSize(), false)
 	}
 
 	// 10000 iterations with random offsets and lengths.
@@ -304,7 +306,8 @@ func benchmarkErasureDecode(data, parity, dataDown, parityDown int, size int64, 
 		if disk == nil {
 			continue
 		}
-		writers[i] = newBitrotWriter(disk, "testbucket", "object", erasure.ShardFileSize(size), DefaultBitrotAlgorithm, erasure.ShardSize())
+		writers[i] = newBitrotWriter(disk, "testbucket", "object",
+			erasure.ShardFileSize(size), DefaultBitrotAlgorithm, erasure.ShardSize(), false)
 	}
 
 	content := make([]byte, size)

--- a/cmd/erasure-encode_test.go
+++ b/cmd/erasure-encode_test.go
@@ -108,7 +108,7 @@ func TestErasureEncode(t *testing.T) {
 			if disk == OfflineDisk {
 				continue
 			}
-			writers[i] = newBitrotWriter(disk, "testbucket", "object", erasure.ShardFileSize(int64(len(data[test.offset:]))), test.algorithm, erasure.ShardSize())
+			writers[i] = newBitrotWriter(disk, "testbucket", "object", erasure.ShardFileSize(int64(len(data[test.offset:]))), test.algorithm, erasure.ShardSize(), false)
 		}
 		n, err := erasure.Encode(context.Background(), bytes.NewReader(data[test.offset:]), writers, buffer, erasure.dataBlocks+1)
 		closeBitrotWriters(writers)
@@ -132,14 +132,14 @@ func TestErasureEncode(t *testing.T) {
 				if disk == nil {
 					continue
 				}
-				writers[i] = newBitrotWriter(disk, "testbucket", "object2", erasure.ShardFileSize(int64(len(data[test.offset:]))), test.algorithm, erasure.ShardSize())
+				writers[i] = newBitrotWriter(disk, "testbucket", "object2", erasure.ShardFileSize(int64(len(data[test.offset:]))), test.algorithm, erasure.ShardSize(), false)
 			}
 			for j := range disks[:test.offDisks] {
 				switch w := writers[j].(type) {
 				case *wholeBitrotWriter:
 					w.disk = badDisk{nil}
 				case *streamingBitrotWriter:
-					w.iow.CloseWithError(errFaultyDisk)
+					w.iow.(*io.PipeWriter).CloseWithError(errFaultyDisk)
 				}
 			}
 			if test.offDisks > 0 {
@@ -196,7 +196,8 @@ func benchmarkErasureEncode(data, parity, dataDown, parityDown int, size int64, 
 				continue
 			}
 			disk.Delete(context.Background(), "testbucket", "object", false)
-			writers[i] = newBitrotWriter(disk, "testbucket", "object", erasure.ShardFileSize(size), DefaultBitrotAlgorithm, erasure.ShardSize())
+			writers[i] = newBitrotWriter(disk, "testbucket", "object",
+				erasure.ShardFileSize(size), DefaultBitrotAlgorithm, erasure.ShardSize(), false)
 		}
 		_, err := erasure.Encode(context.Background(), bytes.NewReader(content), writers, buffer, erasure.dataBlocks+1)
 		closeBitrotWriters(writers)

--- a/cmd/erasure-heal_test.go
+++ b/cmd/erasure-heal_test.go
@@ -87,7 +87,8 @@ func TestErasureHeal(t *testing.T) {
 		buffer := make([]byte, test.blocksize, 2*test.blocksize)
 		writers := make([]io.Writer, len(disks))
 		for i, disk := range disks {
-			writers[i] = newBitrotWriter(disk, "testbucket", "testobject", erasure.ShardFileSize(test.size), test.algorithm, erasure.ShardSize())
+			writers[i] = newBitrotWriter(disk, "testbucket", "testobject",
+				erasure.ShardFileSize(test.size), test.algorithm, erasure.ShardSize(), true)
 		}
 		_, err = erasure.Encode(context.Background(), bytes.NewReader(data), writers, buffer, erasure.dataBlocks+1)
 		closeBitrotWriters(writers)
@@ -130,7 +131,8 @@ func TestErasureHeal(t *testing.T) {
 				continue
 			}
 			os.Remove(pathJoin(disk.String(), "testbucket", "testobject"))
-			staleWriters[i] = newBitrotWriter(disk, "testbucket", "testobject", erasure.ShardFileSize(test.size), test.algorithm, erasure.ShardSize())
+			staleWriters[i] = newBitrotWriter(disk, "testbucket", "testobject",
+				erasure.ShardFileSize(test.size), test.algorithm, erasure.ShardSize(), true)
 		}
 
 		// test case setup is complete - now call Heal()

--- a/cmd/erasure-healing.go
+++ b/cmd/erasure-healing.go
@@ -422,7 +422,8 @@ func (er erasureObjects) healObject(ctx context.Context, bucket string, object s
 					continue
 				}
 				partPath := pathJoin(tmpID, dataDir, fmt.Sprintf("part.%d", partNumber))
-				writers[i] = newBitrotWriter(disk, minioMetaTmpBucket, partPath, tillOffset, DefaultBitrotAlgorithm, erasure.ShardSize())
+				writers[i] = newBitrotWriter(disk, minioMetaTmpBucket, partPath,
+					tillOffset, DefaultBitrotAlgorithm, erasure.ShardSize(), true)
 			}
 			err = erasure.Heal(ctx, readers, writers, partSize)
 			closeBitrotReaders(readers)

--- a/cmd/erasure-multipart.go
+++ b/cmd/erasure-multipart.go
@@ -476,7 +476,8 @@ func (er erasureObjects) PutObjectPart(ctx context.Context, bucket, object, uplo
 		if disk == nil {
 			continue
 		}
-		writers[i] = newBitrotWriter(disk, minioMetaTmpBucket, tmpPartPath, erasure.ShardFileSize(data.Size()), DefaultBitrotAlgorithm, erasure.ShardSize())
+		writers[i] = newBitrotWriter(disk, minioMetaTmpBucket, tmpPartPath,
+			erasure.ShardFileSize(data.Size()), DefaultBitrotAlgorithm, erasure.ShardSize(), false)
 	}
 
 	n, err := erasure.Encode(ctx, data, writers, buffer, writeQuorum)

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -718,7 +718,8 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 		if disk == nil {
 			continue
 		}
-		writers[i] = newBitrotWriter(disk, minioMetaTmpBucket, tempErasureObj, erasure.ShardFileSize(data.Size()), DefaultBitrotAlgorithm, erasure.ShardSize())
+		writers[i] = newBitrotWriter(disk, minioMetaTmpBucket, tempErasureObj,
+			erasure.ShardFileSize(data.Size()), DefaultBitrotAlgorithm, erasure.ShardSize(), false)
 	}
 
 	n, erasureErr := erasure.Encode(ctx, data, writers, buffer, writeQuorum)

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -577,15 +577,6 @@ func (sys *IAMSys) Init(ctx context.Context, objAPI ObjectLayer) {
 	// Hold the lock for migration only.
 	txnLk := objAPI.NewNSLock(minioMetaBucket, minioConfigPrefix+"/iam.lock")
 
-	// Initializing IAM sub-system needs a retry mechanism for
-	// the following reasons:
-	//  - Read quorum is lost just after the initialization
-	//    of the object layer.
-	//  - Write quorum not met when upgrading configuration
-	//    version is needed, migration is needed etc.
-	rquorum := InsufficientReadQuorum{}
-	wquorum := InsufficientWriteQuorum{}
-
 	// allocate dynamic timeout once before the loop
 	iamLockTimeout := newDynamicTimeout(5*time.Second, 3*time.Second)
 
@@ -620,12 +611,7 @@ func (sys *IAMSys) Init(ctx context.Context, objAPI ObjectLayer) {
 		// Migrate IAM configuration, if necessary.
 		if err := sys.doIAMConfigMigration(ctx); err != nil {
 			txnLk.Unlock()
-			if errors.Is(err, errDiskNotFound) ||
-				errors.Is(err, errConfigNotFound) ||
-				errors.Is(err, context.DeadlineExceeded) ||
-				errors.As(err, &rquorum) ||
-				errors.As(err, &wquorum) ||
-				isErrBucketNotFound(err) {
+			if configRetriableErrors(err) {
 				logger.Info("Waiting for all MinIO IAM sub-system to be initialized.. possible cause (%v)", err)
 				continue
 			}
@@ -641,12 +627,7 @@ func (sys *IAMSys) Init(ctx context.Context, objAPI ObjectLayer) {
 
 	for {
 		if err := sys.store.loadAll(ctx, sys); err != nil {
-			if errors.Is(err, errDiskNotFound) ||
-				errors.Is(err, errConfigNotFound) ||
-				errors.Is(err, context.DeadlineExceeded) ||
-				errors.As(err, &rquorum) ||
-				errors.As(err, &wquorum) ||
-				isErrBucketNotFound(err) {
+			if configRetriableErrors(err) {
 				logger.Info("Waiting for all MinIO IAM sub-system to be initialized.. possible cause (%v)", err)
 				time.Sleep(time.Duration(r.Float64() * float64(5*time.Second)))
 				continue

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -237,6 +237,28 @@ func newAllSubsystems() {
 	globalBucketTargetSys = NewBucketTargetSys()
 }
 
+func configRetriableErrors(err error) bool {
+	// Initializing sub-systems needs a retry mechanism for
+	// the following reasons:
+	//  - Read quorum is lost just after the initialization
+	//    of the object layer.
+	//  - Write quorum not met when upgrading configuration
+	//    version is needed, migration is needed etc.
+	rquorum := InsufficientReadQuorum{}
+	wquorum := InsufficientWriteQuorum{}
+
+	// One of these retriable errors shall be retried.
+	return errors.Is(err, errDiskNotFound) ||
+		errors.Is(err, errConfigNotFound) ||
+		errors.Is(err, context.DeadlineExceeded) ||
+		errors.Is(err, errErasureWriteQuorum) ||
+		errors.Is(err, errErasureReadQuorum) ||
+		errors.As(err, &rquorum) ||
+		errors.As(err, &wquorum) ||
+		isErrBucketNotFound(err) ||
+		errors.Is(err, os.ErrDeadlineExceeded)
+}
+
 func initServer(ctx context.Context, newObject ObjectLayer) error {
 	// Once the config is fully loaded, initialize the new object layer.
 	setObjectLayer(newObject)
@@ -251,15 +273,6 @@ func initServer(ctx context.Context, newObject ObjectLayer) error {
 	// ****  WARNING ****
 	// Migrating to encrypted backend should happen before initialization of any
 	// sub-systems, make sure that we do not move the above codeblock elsewhere.
-
-	// Initializing sub-systems needs a retry mechanism for
-	// the following reasons:
-	//  - Read quorum is lost just after the initialization
-	//    of the object layer.
-	//  - Write quorum not met when upgrading configuration
-	//    version is needed, migration is needed etc.
-	rquorum := InsufficientReadQuorum{}
-	wquorum := InsufficientWriteQuorum{}
 
 	r := rand.New(rand.NewSource(time.Now().UnixNano()))
 
@@ -307,15 +320,7 @@ func initServer(ctx context.Context, newObject ObjectLayer) error {
 
 		txnLk.Unlock() // Unlock the transaction lock and allow other nodes to acquire the lock if possible.
 
-		// One of these retriable errors shall be retried.
-		if errors.Is(err, errDiskNotFound) ||
-			errors.Is(err, errConfigNotFound) ||
-			errors.Is(err, context.DeadlineExceeded) ||
-			errors.Is(err, errErasureWriteQuorum) ||
-			errors.Is(err, errErasureReadQuorum) ||
-			errors.As(err, &rquorum) ||
-			errors.As(err, &wquorum) ||
-			isErrBucketNotFound(err) {
+		if configRetriableErrors(err) {
 			logger.Info("Waiting for all MinIO sub-systems to be initialized.. possible cause (%v)", err)
 			time.Sleep(time.Duration(r.Float64() * float64(5*time.Second)))
 			continue
@@ -333,8 +338,6 @@ func initAllSubsystems(ctx context.Context, newObject ObjectLayer) (err error) {
 	// you want to add extra context to your error. This
 	// ensures top level retry works accordingly.
 	// List buckets to heal, and be re-used for loading configs.
-	rquorum := InsufficientReadQuorum{}
-	wquorum := InsufficientWriteQuorum{}
 
 	buckets, err := newObject.ListBuckets(ctx)
 	if err != nil {
@@ -368,14 +371,7 @@ func initAllSubsystems(ctx context.Context, newObject ObjectLayer) (err error) {
 
 	// Initialize config system.
 	if err = globalConfigSys.Init(newObject); err != nil {
-		if errors.Is(err, errDiskNotFound) ||
-			errors.Is(err, errConfigNotFound) ||
-			errors.Is(err, context.DeadlineExceeded) ||
-			errors.Is(err, errErasureWriteQuorum) ||
-			errors.Is(err, errErasureReadQuorum) ||
-			errors.As(err, &rquorum) ||
-			errors.As(err, &wquorum) ||
-			isErrBucketNotFound(err) {
+		if configRetriableErrors(err) {
 			return fmt.Errorf("Unable to initialize config system: %w", err)
 		}
 		// Any other config errors we simply print a message and proceed forward.

--- a/cmd/xl-storage-disk-id-check.go
+++ b/cmd/xl-storage-disk-id-check.go
@@ -62,7 +62,7 @@ const (
 
 // Detects change in underlying disk.
 type xlStorageDiskIDCheck struct {
-	storage *xlStorage
+	storage StorageAPI
 	diskID  string
 
 	apiCalls     [metricLast]uint64

--- a/cmd/xl-storage.go
+++ b/cmd/xl-storage.go
@@ -71,6 +71,13 @@ const (
 	xlStorageFormatFile = "xl.meta"
 )
 
+var alignedBuf []byte
+
+func init() {
+	alignedBuf = disk.AlignedBlock(4096)
+	_, _ = rand.Read(alignedBuf)
+}
+
 // isValidVolname verifies a volname name in accordance with object
 // layer requirements.
 func isValidVolname(volname string) bool {
@@ -282,10 +289,17 @@ func newXLStorage(ep Endpoint) (*xlStorage, error) {
 	var rnd [8]byte
 	_, _ = rand.Read(rnd[:])
 	tmpFile := ".writable-check-" + hex.EncodeToString(rnd[:]) + ".tmp"
-	if err = p.CreateFile(GlobalContext, minioMetaTmpBucket, tmpFile, 1, strings.NewReader("0")); err != nil {
+	filePath := pathJoin(p.diskPath, minioMetaTmpBucket, tmpFile)
+	w, err := disk.OpenFileDirectIO(filePath, os.O_CREATE|os.O_WRONLY|os.O_EXCL, 0666)
+	if err != nil {
 		return p, err
 	}
-	defer os.Remove(pathJoin(p.diskPath, minioMetaTmpBucket, tmpFile))
+	if _, err = w.Write(alignedBuf[:]); err != nil {
+		w.Close()
+		return p, err
+	}
+	w.Close()
+	defer os.Remove(filePath)
 
 	volumeDir, err := p.getVolDir(minioMetaTmpBucket)
 	if err != nil {
@@ -294,7 +308,7 @@ func newXLStorage(ep Endpoint) (*xlStorage, error) {
 
 	// Check if backend is readable, and optionally supports O_DIRECT.
 	if _, err = p.readAllData(volumeDir, pathJoin(volumeDir, tmpFile), true); err != nil {
-		if err != errUnsupportedDisk {
+		if !errors.Is(err, errUnsupportedDisk) {
 			return p, err
 		}
 		// error is unsupported disk, turn-off directIO for reads
@@ -1432,7 +1446,7 @@ func (s *xlStorage) CreateFile(ctx context.Context, volume, path string, fileSiz
 		return errInvalidArgument
 	}
 
-	if fileSize <= smallFileThreshold {
+	if fileSize >= 0 && fileSize <= smallFileThreshold {
 		// For streams smaller than 128KiB we simply write them as O_DSYNC (fdatasync)
 		// and not O_DIRECT to avoid the complexities of aligned I/O.
 		w, err := s.openFile(volume, path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC)
@@ -1483,9 +1497,9 @@ func (s *xlStorage) CreateFile(ctx context.Context, volume, path string, fileSiz
 		return err
 	}
 
-	if written < fileSize {
+	if written < fileSize && fileSize >= 0 {
 		return errLessData
-	} else if written > fileSize {
+	} else if written > fileSize && fileSize >= 0 {
 		return errMoreData
 	}
 

--- a/cmd/xl-storage_test.go
+++ b/cmd/xl-storage_test.go
@@ -1730,7 +1730,7 @@ func TestXLStorageVerifyFile(t *testing.T) {
 	// 4) Streaming bitrot check on corrupted file
 
 	// create xlStorage test setup
-	xlStorage, path, err := newXLStorageTestSetup()
+	storage, path, err := newXLStorageTestSetup()
 	if err != nil {
 		t.Fatalf("Unable to create xlStorage test setup, %s", err)
 	}
@@ -1738,7 +1738,7 @@ func TestXLStorageVerifyFile(t *testing.T) {
 
 	volName := "testvol"
 	fileName := "testfile"
-	if err := xlStorage.MakeVol(context.Background(), volName); err != nil {
+	if err := storage.MakeVol(context.Background(), volName); err != nil {
 		t.Fatal(err)
 	}
 
@@ -1752,29 +1752,29 @@ func TestXLStorageVerifyFile(t *testing.T) {
 	h := algo.New()
 	h.Write(data)
 	hashBytes := h.Sum(nil)
-	if err := xlStorage.WriteAll(context.Background(), volName, fileName, data); err != nil {
+	if err := storage.WriteAll(context.Background(), volName, fileName, data); err != nil {
 		t.Fatal(err)
 	}
-	if err := xlStorage.storage.bitrotVerify(pathJoin(path, volName, fileName), size, algo, hashBytes, 0); err != nil {
+	if err := storage.storage.(*xlStorage).bitrotVerify(pathJoin(path, volName, fileName), size, algo, hashBytes, 0); err != nil {
 		t.Fatal(err)
 	}
 
 	// 2) Whole-file bitrot check on corrupted file
-	if err := xlStorage.AppendFile(context.Background(), volName, fileName, []byte("a")); err != nil {
+	if err := storage.AppendFile(context.Background(), volName, fileName, []byte("a")); err != nil {
 		t.Fatal(err)
 	}
 
 	// Check if VerifyFile reports the incorrect file length (the correct length is `size+1`)
-	if err := xlStorage.storage.bitrotVerify(pathJoin(path, volName, fileName), size, algo, hashBytes, 0); err == nil {
+	if err := storage.storage.(*xlStorage).bitrotVerify(pathJoin(path, volName, fileName), size, algo, hashBytes, 0); err == nil {
 		t.Fatal("expected to fail bitrot check")
 	}
 
 	// Check if bitrot fails
-	if err := xlStorage.storage.bitrotVerify(pathJoin(path, volName, fileName), size+1, algo, hashBytes, 0); err == nil {
+	if err := storage.storage.(*xlStorage).bitrotVerify(pathJoin(path, volName, fileName), size+1, algo, hashBytes, 0); err == nil {
 		t.Fatal("expected to fail bitrot check")
 	}
 
-	if err := xlStorage.Delete(context.Background(), volName, fileName, false); err != nil {
+	if err := storage.Delete(context.Background(), volName, fileName, false); err != nil {
 		t.Fatal(err)
 	}
 
@@ -1782,7 +1782,7 @@ func TestXLStorageVerifyFile(t *testing.T) {
 	algo = HighwayHash256S
 	shardSize := int64(1024 * 1024)
 	shard := make([]byte, shardSize)
-	w := newStreamingBitrotWriter(xlStorage, volName, fileName, size, algo, shardSize)
+	w := newStreamingBitrotWriter(storage, volName, fileName, size, algo, shardSize, false)
 	reader := bytes.NewReader(data)
 	for {
 		// Using io.Copy instead of this loop will not work for us as io.Copy
@@ -1798,13 +1798,13 @@ func TestXLStorageVerifyFile(t *testing.T) {
 		}
 		t.Fatal(err)
 	}
-	w.Close()
-	if err := xlStorage.storage.bitrotVerify(pathJoin(path, volName, fileName), size, algo, nil, shardSize); err != nil {
+	w.(io.Closer).Close()
+	if err := storage.storage.(*xlStorage).bitrotVerify(pathJoin(path, volName, fileName), size, algo, nil, shardSize); err != nil {
 		t.Fatal(err)
 	}
 
 	// 4) Streaming bitrot check on corrupted file
-	filePath := pathJoin(xlStorage.String(), volName, fileName)
+	filePath := pathJoin(storage.String(), volName, fileName)
 	f, err := os.OpenFile(filePath, os.O_WRONLY|os.O_SYNC, 0644)
 	if err != nil {
 		t.Fatal(err)
@@ -1814,10 +1814,10 @@ func TestXLStorageVerifyFile(t *testing.T) {
 		t.Fatal(err)
 	}
 	f.Close()
-	if err := xlStorage.storage.bitrotVerify(pathJoin(path, volName, fileName), size, algo, nil, shardSize); err == nil {
+	if err := storage.storage.(*xlStorage).bitrotVerify(pathJoin(path, volName, fileName), size, algo, nil, shardSize); err == nil {
 		t.Fatal("expected to fail bitrot check")
 	}
-	if err := xlStorage.storage.bitrotVerify(pathJoin(path, volName, fileName), size+1, algo, nil, shardSize); err == nil {
+	if err := storage.storage.(*xlStorage).bitrotVerify(pathJoin(path, volName, fileName), size+1, algo, nil, shardSize); err == nil {
 		t.Fatal("expected to fail bitrot check")
 	}
 }

--- a/pkg/ioutil/ioutil.go
+++ b/pkg/ioutil/ioutil.go
@@ -20,8 +20,10 @@ package ioutil
 
 import (
 	"bytes"
+	"context"
 	"io"
 	"os"
+	"time"
 
 	"github.com/minio/minio/pkg/disk"
 )
@@ -62,6 +64,56 @@ func (w *WriteOnCloser) HasWritten() bool { return w.hasWritten }
 // WriteOnClose takes an io.Writer and returns an ioutil.WriteOnCloser.
 func WriteOnClose(w io.Writer) *WriteOnCloser {
 	return &WriteOnCloser{w, false}
+}
+
+type ioret struct {
+	n   int
+	err error
+}
+
+// DeadlineWriter deadline writer with context
+type DeadlineWriter struct {
+	io.WriteCloser
+	timeout time.Duration
+	err     error
+}
+
+// NewDeadlineWriter wraps a writer to make it respect given deadline
+// value per Write(). If there is a blocking write, the returned Writer
+// will return whenever the timer hits (the return values are n=0
+// and err=context.Canceled.)
+func NewDeadlineWriter(w io.WriteCloser, timeout time.Duration) io.WriteCloser {
+	return &DeadlineWriter{WriteCloser: w, timeout: timeout}
+}
+
+func (w *DeadlineWriter) Write(buf []byte) (int, error) {
+	if w.err != nil {
+		return 0, w.err
+	}
+
+	c := make(chan ioret, 1)
+	t := time.NewTimer(w.timeout)
+	defer t.Stop()
+
+	go func() {
+		n, err := w.WriteCloser.Write(buf)
+		c <- ioret{n, err}
+		close(c)
+	}()
+
+	select {
+	case r := <-c:
+		w.err = r.err
+		return r.n, r.err
+	case <-t.C:
+		w.err = context.Canceled
+		return 0, context.Canceled
+	}
+}
+
+// Close closer interface to close the underlying closer
+func (w *DeadlineWriter) Close() error {
+	return w.WriteCloser.Close()
 }
 
 // LimitWriter implements io.WriteCloser.

--- a/pkg/ioutil/ioutil_test.go
+++ b/pkg/ioutil/ioutil_test.go
@@ -18,11 +18,48 @@ package ioutil
 
 import (
 	"bytes"
+	"context"
 	"io"
 	goioutil "io/ioutil"
 	"os"
 	"testing"
+	"time"
 )
+
+type sleepWriter struct {
+	timeout time.Duration
+}
+
+func (w *sleepWriter) Write(p []byte) (n int, err error) {
+	time.Sleep(w.timeout)
+	return len(p), nil
+}
+
+func (w *sleepWriter) Close() error {
+	return nil
+}
+
+func TestDeadlineWriter(t *testing.T) {
+	w := NewDeadlineWriter(&sleepWriter{timeout: 500 * time.Millisecond}, 450*time.Millisecond)
+	_, err := w.Write([]byte("1"))
+	w.Close()
+	if err != context.Canceled {
+		t.Error("DeadlineWriter shouldn't be successful - should return context.Canceled")
+	}
+	_, err = w.Write([]byte("1"))
+	if err != context.Canceled {
+		t.Error("DeadlineWriter shouldn't be successful - should return context.Canceled")
+	}
+	w = NewDeadlineWriter(&sleepWriter{timeout: 500 * time.Millisecond}, 600*time.Millisecond)
+	n, err := w.Write([]byte("abcd"))
+	w.Close()
+	if err != nil {
+		t.Errorf("DeadlineWriter should succeed but failed with %s", err)
+	}
+	if n != 4 {
+		t.Errorf("DeadlineWriter should succeed but should have only written 0 bytes, but returned %d instead", n)
+	}
+}
 
 func TestCloseOnWriter(t *testing.T) {
 	writer := WriteOnClose(goioutil.Discard)

